src/main/java/core/basesyntax/service/impl/ProductCounterServiceImpl.java

```
@@ -0,0 +1,58 @@
+import java.util.List;
+
+public class ProductCounterServiceImpl implements ProductCounterService {
+    private final ActionService actionService;
+
+    public ProductCounterServiceImpl(ActionService actionService) {
+        if (actionService == null) {
+            throw new ActionServiceIsNullException("ActionService can't be null");
+        }
+        this.actionService = actionService;
+    }
+
+    @Override
+    public void countTheProducts(List<FruitTransaction> fruitTransactions, Storage storage) {
+        if (fruitTransactions == null) {
+            throw new ReportIsNullException("Report can't be null");
+        }
+        if (storage == null) {
+            throw new StorageIsNullException("Storage can't be null");
+        }
+        if (fruitTransactions.size() <= 1) {
+            return;
+        }
+        for (FruitTransaction fruitTransaction : fruitTransactions) {
+            String productName = fruitTransaction.getProductName();
+            int quantity = fruitTransaction.getQuantity();
+            String actionCode = fruitTransaction.getActionCode();
+            if (actionCode == null || productName == null) {
+                throw new IncorrectFormatOfDataException("ActionCode"
+                        + " and ProductName can't be null");
+            }
+            if (actionCode.isEmpty() ||  productName.isEmpty()) {
+                throw new IncorrectFormatOfDataException(
+                        "Fruit transaction: action/product must not be blank -> " +
+                                fruitTransaction
+                );
+            }
+            if (quantity < 0) {
+                throw new QuantityLessThanNullException("Quantity must be "
+                        + "greater than or equal to 0: " + quantity);
+            }
+            ActionTypeService action = actionService.getStrategy(actionCode);
+            action.applyTheQuantity(storage, productName, quantity);
```

This method validates that the `fruitTransactions` list itself is not null, but it does not validate that individual elements inside the list are non-null before dereferencing them. The task requires defensive handling of null parameters/inputs: "Handle null parameters defensively (throw informative RuntimeException)." Currently the for-loop assumes each `fruitTransaction` is non-null and calls `fruitTransaction.getProductName()` etc., which will throw NPE if a null element appears. Add a null-element check (and/or validate the list contents when parsing) and throw an informative RuntimeException if an element is null so the service fails fast with a clear message.